scripts/exact_1d_bayesian_filtering_utils.py

```diff
@@ -0,0 +1,384 @@
+from typing import Callable
+import functools
+
+import jax
+from jax import numpy as jnp
+from jax import scipy as jsp
+from jax import random
+import matplotlib.pyplot as plt
+
+
+plt.rcParams.update({"font.size": 16})
+
+
+def generate_x_true(rng_key, K, x0_rvs: Callable, v_rvs: Callable, f: Callable):
+    rng_keys = random.split(rng_key, num=2)
+    x0 = x0_rvs(rng_keys[0], shape=())
+    x_true = jnp.zeros(shape=(K + 1,))
+    x_true = x_true.at[0].set(x0)
+    v = v_rvs(rng_keys[1], shape=x_true.shape)
+    for k in range(1, K + 1):
+        x_true = x_true.at[k].set(f(x_true[k - 1], v[k - 1], k=k - 1))
+    return jnp.array(x_true)
+
+
+def generate_y(rng_key, x_true, e_rvs: Callable, h: Callable):
+    shape = x_true.shape
+    e = e_rvs(rng_key, shape=shape)
+    y = h(x_true, e)
+    y = y.at[0].set(jnp.inf)
+    return y
+
+
+def x_pdf(x_new, x, k, v_pdf, f):
+    v = x_new - f(x=x, v=0, k=k)
+    return v_pdf(x - v)
+
+
+def y_likelihood(y, x, e_pdf, h):
+    e = h(x=x, e=0)
+    return e_pdf(y - e)
+
+
+def inversion_sampling(rng_key, x_grid, px_grid, num_samples):
+    rng_keys = random.split(rng_key, num=2)
+    u = random.uniform(rng_keys[0], shape=(num_samples, 1))
+    delta = x_grid[1] - x_grid[0]
+    noise = random.uniform(
+        rng_keys[1], minval=-delta / 2, maxval=delta / 2, shape=(num_samples,)
+    )
+    # It only works for sufficient dense uniformly spaced grid
+    point_mass = px_grid
+
+    cdf = jnp.cumsum(point_mass)
+
+    bound_cdf = jnp.where(cdf < u, cdf, 0)
+    idx = jnp.argmax(bound_cdf, axis=1)
+    x = x_grid[idx]
+    return x + noise
+
+
+def kde(x_grid, x, kernel_variance):
+    delta = x_grid[1] - x_grid[0]
+
+    px = jax.vmap(functools.partial(jsp.stats.norm.pdf, loc=x, scale=kernel_variance))(
+        x_grid
+    )
```

Comment:
Better if you divide this in two/three steps (lines of code): the first one defines the partial function, then you define the vmap, and then you apply the vmap to get `px`. 
Also, I'm not sure if you need to vmap this function and apply a partial to it. You should be able to do that directly from jax.scipy.stats.